src/mds/Server.cc

```diff
@@ -2334,18 +2334,20 @@ void Server::set_trace_dist(const ref_t<MClientReply> &reply,
     encode(dn->get_name(), bl);
 
     int lease_mask = 0;
-    CDentry::linkage_t *dnl = dn->get_linkage(mdr->get_client(), mdr);
-    if (dnl->is_primary()) {
-      ceph_assert(dnl->get_inode() == in);
-      lease_mask = CEPH_LEASE_PRIMARY_LINK;
-    } else {
-      if (dnl->is_remote())
-	ceph_assert(dnl->get_remote_ino() == in->ino());
-      else
-	ceph_assert(!in);
+    if (!dn->is_projected() || dn->use_projected(mdr->get_client(), mdr)) {
+      CDentry::linkage_t *dnl = dn->get_linkage(mdr->get_client(), mdr);
+      if (dnl->is_primary()) {
+        ceph_assert(dnl->get_inode() == in);
+        lease_mask = CEPH_LEASE_PRIMARY_LINK;
+      } else {
+        if (dnl->is_remote())
+          ceph_assert(dnl->get_remote_ino() == in->ino());
+        else
+          ceph_assert(!in);
+      }
+      mds->locker->issue_client_lease(dn, mdr, lease_mask, now, bl);
+      dout(20) << "set_trace_dist added dn   " << snapid << " " << *dn << dendl;
     }
-    mds->locker->issue_client_lease(dn, mdr, lease_mask, now, bl);
-    dout(20) << "set_trace_dist added dn   " << snapid << " " << *dn << dendl;
```

Comment:
I think @ukernel  was consolidating the logic of issuing a null lease in `Locker::issue_client_lease` if a lease cannot be issued. This changeset does not issue such a null lease. I'm concerned that may lead to bugs.

I think the issue is that the assert is simply wrong. The dentry's durable linkage may not match the projected linkage for all clients which may hit this method. It should only match if `dn->lock.can_lease(client)` is true? I think the assert should be moved to `Locker::issue_client_lease`. The logic for creating the lease_mask also looks like it can be moved there too (code maint. only).

Reply:
> I think @ukernel was consolidating the logic of issuing a null lease in `Locker::issue_client_lease` if a lease cannot be issued. This changeset does not issue such a null lease. I'm concerned that may lead to bugs.

Well, clients can handle mds trace replies without a lease, isn't it? I'm not sure if sending a trace reply without a lease would be buggy. Sending a null lease does let the client not query the mds though.

> 
> I think the issue is that the assert is simply wrong. The dentry's durable linkage inode may not match the projected linkage inode for all clients which may hit this method. It should only match if `dn->lock.can_lease(client)` is true? I think the assert should be moved to `Locker::issue_client_lease`. The logic for creating the lease_mask also looks like it can be moved there too (code maint. only).

OK. This sounds fine. Checked the state transition for `can_lease` vs `can_read_projected` - checking `can_lease` does seems to be much more comprehensive check on when leases can be granted to clients.